internal/gcs/container.go

```diff
@@ -0,0 +1,171 @@
+	}
+	if _, ok := gc.notifyChs[cid]; ok {
+		return fmt.Errorf("container %s already exists", cid)
+	}
+	gc.notifyChs[cid] = ch
+	return nil
+}
+
+func (gc *GuestConnection) notify(ntf *containerNotification) error {
+	cid := ntf.ContainerID
+	gc.mu.Lock()
+	ch := gc.notifyChs[cid]
+	delete(gc.notifyChs, cid)
+	gc.mu.Unlock()
+	if ch == nil {
+		return fmt.Errorf("container %s not found", cid)
+	}
+	logrus.WithField(logfields.ContainerID, cid).Info("container terminated in guest")
+	close(ch)
+	return nil
+}
+
+func (gc *GuestConnection) clearNotifies() {
+	gc.mu.Lock()
+	chs := gc.notifyChs
+	gc.notifyChs = nil
+	gc.mu.Unlock()
+	for _, ch := range chs {
+		close(ch)
+	}
+}
+
+type Container struct {
+	gc        *GuestConnection
+	id        string
+	notifyCh  chan struct{}
+	closeCh   chan struct{}
+	closeOnce sync.Once
+}
+
+func (gc *GuestConnection) CreateContainer(ctx context.Context, cid string, cfg interface{}) (*Container, error) {
+	c := &Container{
+		gc:       gc,
+		id:       cid,
+		notifyCh: make(chan struct{}),
+		closeCh:  make(chan struct{}),
+	}
+	err := gc.requestNotify(cid, c.notifyCh)
+	if err != nil {
+		return nil, err
+	}
+	req := containerCreate{
+		requestBase:     makeRequest(cid),
+		ContainerConfig: anyInString{cfg},
+	}
+	var resp containerCreateResponse
+	err = gc.brdg.RPC(ctx, rpcCreate, &req, &resp, false)
+	if err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+func (c *Container) Close() error {
+	c.closeOnce.Do(func() {
+		close(c.closeCh)
+	})
+	return nil
+}
+
+func (c *Container) CreateProcess(config interface{}) (cow.Process, error) {
+	return c.gc.exec(context.TODO(), c.id, config)
+}
+
+func (c *Container) ID() string {
+	return c.id
+}
+
+func (c *Container) Modify(config interface{}) (err error) {
+	req := containerModifySettings{
+		requestBase: makeRequest(c.id),
+		Request:     config,
+	}
+	var resp responseBase
+	return c.gc.brdg.RPC(context.TODO(), rpcModifySettings, &req, &resp, false)
+}
+
+func (c *Container) Properties(types ...schema1.PropertyType) (_ *schema1.ContainerProperties, err error) {
+	req := containerGetProperties{
+		requestBase: makeRequest(c.id),
+		Query:       "",
+	}
+	var resp containerGetPropertiesResponse
+	err = c.gc.brdg.RPC(context.TODO(), rpcGetProperties, &req, &resp, true)
+	if err != nil {
+		return nil, err
+	}
+	var props schema1.ContainerProperties
+	err = json.Unmarshal(resp.Properties, &props)
+	if err != nil {
+		return nil, err
+	}
+	return nil, nil
+}
+
+func (c *Container) Start() error {
+	req := makeRequest(c.id)
+	var resp responseBase
+	return c.gc.brdg.RPC(context.TODO(), rpcStart, &req, &resp, false)
+}
+
+func (c *Container) shutdown(ctx context.Context, proc rpcProc) error {
+	req := makeRequest(c.id)
+	var resp responseBase
+	err := c.gc.brdg.RPC(ctx, proc, &req, &resp, true)
+	if err != nil {
+		if uint32(resp.Result) != hrComputeSystemDoesNotExist {
+			return err
+		}
+		select {
+		case <-c.notifyCh:
```

Comment:
I don't think you want this right?  This read could read the shutdown notification and `Wait()` would then result in `container closed` rather than an expected exit. I think we just want to log the "ignoring missing container" message right?

Reply:
Oh. Actually `Wait()` would block forever because the `Wait()` routine wouldn't unblock so the caller would call `Close()`

Reply:
What do you mean? `notifyCh` is closed when either we get the notification or the bridge is terminated. `closeCh` is closed when the caller calls `Close`.

Reply:
Just looked at this again. My interpretation of what the "notify channel" was used for was to send the notify message. I see that its just used as a wait block so this is fine.